### PR TITLE
feat: LTI 1.3 reusable configuration

### PR DESCRIPTION
**Pull request by kuipumu**
## Description

This PR adds compatibility for LTI 1.3 external configurations to the LTI consumer XBlock. This change allow the XBlock to use values from the external LTI configuration when enabled and setup. This PR also includes a fix that adds validation to the external_config field on the LTI consumer XBlock and also improves the behavior of the JS that handles the visibility of configuration fields on the LTI consumer XBlock.

## Type of Change

- [x] Modify `urlpatterns` variable on `plugins/urls.py` module.
- [x] Modify `get_lti_1p3_launch_info` function on `api.py` module.
- [x] Modify `LtiConsumerXBlock` `validate_field_data` method.
- [x] Modify `LtiConsumerXBlock` `_get_lti_block_launch_handler` method. 
- [x] Modify `LtiConsumerXBlock` `_get_lti_1p3_launch_url` method. 
- [x] Modify `LtiConfiguration` `clean` method.
- [x] Modify `LtiConfiguration` `get_lti_advantage_ags_mode` method.
- [x] Modify `LtiConfiguration` `get_lti_advantage_deep_linking_enabled` method.
- [x] Modify `LtiConfiguration` `get_lti_advantage_deep_linking_launch_url` method.
- [x] Modify `LtiConfiguration` `get_lti_advantage_nrps_enabled` method.
- [x] Modify `LtiConfiguration` `get_lti_1p3_redirect_uris` method.
- [x] Modify `LtiConfiguration` `_get_lti_1p3_consumer` method.
- [x] Modify `xblock_studio_view.js`.
- [x] Modify `public_keyset_endpoint` function on `plugin/views.py` module.
- [x] Modify `access_token_endpoint` function on `plugin/views.py` module.
- [x] Include tests for all added and modified code.

## Testing

1. Install and configure openedx-ltistore (follow the setup steps from: https://github.com/openedx/xblock-lti-consumer/pull/239 and use the related PR: https://github.com/open-craft/openedx-ltistore/pull/2)
2. Create an IMS reference tool: https://lti-ri.imsglobal.org/lti/tools
3. Go to http://localhost:18000/admin/lti_store/externallticonfiguration/ and create a new external LTI 1.3 configuration.
4. Fill the "LTI 1.3 Private Key",  "LTI 1.3 Client ID" and "LTI 1.3 Tool Keyset URL" or "LTI 1.3 Tool Public Key" field.
5. Create a course and add a LTI 1.3 consumer XBlock.
6. Set the consumer XBlock to user an external configuration and set the reusable configuration ID (follow the steps on how to get the filter key from: https://github.com/openedx/xblock-lti-consumer/pull/239) 
7. Copy the keyset and access token URL from the XBlock preview back to the IMS reference tool previously created. 
8. Go to the course a launch the XBlock.
9. The LTI 1.3 launch should be valid.
10. Check that the XBlock doesn't allow you to set the external configuration type without adding an external configuration ID.
11. Check that the XBlock doesn't allow you to set an invalid external configuration ID format.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @kuipumu!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by mphilbrick211** [comment]
Hi @kuipumu! Is this all set for review/merge?

**Comment by Zacharis278** [comment]
@kuipumu thank you! I'll have more opportunity to dig into these PR's more deeply next week. That said this looks really promising. I could have sworn there was a catch to adopting this implementation for 1.3 but this seems pretty straightforward.

**Comment by kuipumu** [comment]
@Zacharis278 I only know that the OpenCraft team had a plan to decouple the LTI configuration model, that will meant this implementation will need to change to adjust to those changes, right now this implementation depends on having the keyset and access token URLs being on the external app.

@mphilbrick211 Yes, this PR is ready for review/merge.

**Comment by mphilbrick211** [comment]
@Zacharis278 - are you able to merge this if it's good-to-go? I believe this repo is owned by the Cosmonauts, but I'm not sure who (aside from Chris H.) is on the team.

**Comment by mphilbrick211** [comment]
Thanks, @Zacharis278! Putting updates like that in here is super helpful. I might check in again, but if you're still not ready, that's OK. Having an ETA on this, or providing another update would also help if possible.

**Comment by Zacharis278** [comment]
@kuipumu 

This all looks good to me and would fit wiht the long term plan to decouple configuration/placement from what I can tell. Maybe a little more refactoring with merging this first to move these fields back out of configuration and into placement but I don't think that should hold us up.

Brief question on the configurable deployment id. Based on our current model there's no way to have multiple deployment id's for a configuration. How would this get used? Is this just forward thinking for future development of multi-tenancy?

I'm looking through the other PR's and plan to manually test each of them. Should get through that by tomorrow

**Comment by Zacharis278** [comment]
@tecoholic are you the right person to ping on the OpenCraft side to take a look at this? Would like to get your thoughts before merging.

**Comment by kuipumu** [comment]
@Zacharis278 This was included to make external configs be multi-tenancy compatible. I included the possibility to use a deployment ID set on the external config, I added a field `lti_1p3_deployment_id` on the external config, if the field is not set, the default deployment ID of '1' is used. According to what was already commented on the **LtiAgsLineItem** model, I also included the `oidc_url`, `client_id` and `deployment_id` fields to be able isolate each lineitem per deployment. So far I haven't found any issues on this implementation but I might be wrong. Is there any detail I'm missing regarding the use of a different deployment ID?

**Comment by Zacharis278** [comment]
Think I got it. My understanding of how this would work is right now, setting deployment id doesn't really do anything other than keeping us single tenant with a different deployment id, right? Since deployment id is on the configuration it's always a 1:1 relationship. I'm wondering if the configurable deployment id is adding any value. Please correct me if I'm missing a good use case for this.

My concern is that this would be something we need to refactor when we do build the modeling for multi-tenancy. If existing configurations are already using custom deployment ids that could become messy to migrate.

**Comment by kuipumu** [comment]
@Zacharis278 I think there is some confusion regarding the deployment_id, there are a couple of diagrams that more clearly represent the role of the deployment_id value on the LTI tool deployment: https://www.imsglobal.org/spec/lti/v1p3#multi-tenant-tool-registered-once-deployed-multiple-times. The role of the deployment ID is mainly on the tool deployment, the idea is that you can share the same client ID, public key on a tool while creating a distinction between different deployments with the deployment ID, this means that you would just need to deploy the tool once and reuse this same security contract, correct me if I'm wrong in any of this. I think multi-tenancy is already working by just enabling the use of a different deployment ID. There is certain details, for example with AGS, each score and lineitem should be isolated per deployment, this is addressed in here, I don't know of any other resource that's shared between tool and platform that needs to be isolated per deployment, if any that should be addressed, but I didn't notice anything else on that regards that needs to be addressed.

**Comment by tecoholic** [comment]
@Zacharis278 Hi, I think @Agrendalath is going to review a couple of PRs related to LTI. I think he might be the better person to handle this here.

**Comment by Zacharis278** [comment]
Sorry for any confusion, I understand how it works according to the spec but my questions are specifically to the fact that Open edX does not have a way to configure multiple deployments to a single configuration yet. If we do build support for this it would likely be some model separate from the tool configuration itself (probably an 'account' concept). I just want to make sure we're designing in a way that keeps that path open. Our upcoming goals around building a marketplace will eventually need this.

I'd just like to understand how deployment id on configuration as it exists here could be used right now. If there's a good use case, than yeah we should add it. I wouldn't want to get too hung up on future designs if it would block adding something useful right now.

**Comment by kuipumu** [comment]
@Zacharis278 OK I wasn't aware about any of this, right know the use case for this is just to be able to set a deployment ID different from the default "1", so we can for example have a tool that will accept multiple XBlocks using the same client_id but with a different deployment ID so we can differentiate between different deployments of this configuration.

**Review by Agrendalath — CHANGES_REQUESTED** [review]
@kuipumu, thank you for this contribution, and sorry for the delay in the review. I left some comments here and [in the openedx-ltistore PR](https://github.com/open-craft/openedx-ltistore/pull/2#discussion_r1302187024).

File: lti_consumer/models.py
Comment (kuipumu) [reply lti_consumer/models.py:562]: We introduce it so course authors can override the OIDC URL, launch URL or deep linking launch URL, without the need to create a new configuration. We could add also the DB values as a fallback, I didn't wanted to introduce to much options on this fallback but also having the fallback on the DB might be useful. What do you suggest this be documented on?, I just thought about adding a code comment to describe this behavior but I agree it might be better to have it documented.

File: lti_consumer/models.py
Comment (kuipumu) [reply lti_consumer/models.py:683]: A single LtiConfiguration entity can only have one set of oidc_url, client_id, and deployment_id values set on the DB and XBlock, but once we use external configurations, the oidc_url, client_id and deployment_id values on the external configuration could change, this will mean that the gradebook won't change even if the oidc_url, client_id and deployment_id values changed for the LtiConfiguration. Since each lineitem should only expose gradebook information directly tied to the tool deployment, I modified the LtiAgsLineItem model to include the oidc_url, client_id and deployment_id so we are querying the LtiAgsLineItem on LtiAgsLineItemViewset not only by the lti_configuration but also by the values currently being used by the LtiConfiguration. I might be wrong about this, so please correct me if I'm wrong.

Also, you are correct, this fields should be filled in the migration, otherwise, this change wouldn't be backward compatible if applied.

We noticed the possibility of using deployment IDs while using an external configuration, I can separate anything related to implementing the use of multi-tenancy with external configurations on a separate PR to avoid increasing the complexity of the initial implementation of external configurations for LTI 1.3. What do you think if I add an ADR on how to set external configurations for LTI 1.3 and another ADR about external configurations multi-tenancy on this other PR I will open?.

**Comment by kuipumu** [comment]
@Agrendalath Sorry for my delay on the response, I responded a few of your comments here and in [in the openedx-ltistore PR](https://github.com/open-craft/openedx-ltistore/pull/2#discussion_r1311047162). I applied a few of your suggestions to the code, Thanks!.

File: lti_consumer/models.py
Comment (Agrendalath) [reply lti_consumer/models.py]: @kuipumu , ah, right, we use them as overrides, not fallbacks.

Should we also override other values (like `client_id`, `rsa_key`) through the `LtiConfiguration`? _Quick context: I'm talking only about the `LtiConfiguration or ExternalConfiguration` condition since these values cannot be easily modified (or are not stored) within the XBlock fields._

I believe that the most discoverable solution would be adding this information directly to the help text of each XBlock/DB field that can be overridden this way. Something like _"If the configuration is retrieved from an external store, this value overrides the one specified in the external configuration."_, but with an indication that this is not applicable when you use the `CONFIG_ON_DB` option.

@Zacharis278, what do you think?

File: lti_consumer/models.py
Comment (Agrendalath) [reply lti_consumer/models.py:756]: @kuipumu, yes, please move the multi-tenancy-related things to a separate PR. Once we merge the reusable LTI 1.3 configs, catching potential regressions will be much easier.

> What do you think if I add an ADR on how to set external configurations for LTI 1.3 and another ADR about external configurations multi-tenancy on this other PR I will open?.

Sounds reasonable to me. Do we have a rough timeline for this separate PR? It would be good to add the LTI 1.3 external configurations ADR before the Quince release.

File: lti_consumer/models.py
Comment (Zacharis278) [reply lti_consumer/models.py]: I think an issue we run into with the override functionality is right now the configuration options are exclusive. You pick either xblock, database, or external. If you choose external it's ambiguous what the next most important configuration location would be. I'd agree making this not applicable `CONFIG_ON_DB` is probably the simplest short term solution around that. Another thing is that this would differ from the 1.1 implementation. If we're going to add an override feature its behavior should be consistent across LTI versions.

I'm leaning towards suggesting we break override behavior out into it's own PR and instead focus on parity with the existing 1.1 solution in here. Taking time to discuss/consider the full list of fields that are potentially overridable at the block level and how we visually communicate to course teams how it would work might hold up the core behavior of this PR.

Comment (kuipumu) [reply lti_consumer/models.py]: @Zacharis278 I agree that we should make this PR only about making LTI 1.3 external configuration 1:1 with the LTI 1.1 implementation and move other specific features (multi-tenancy, field override) into separate PRs. I already removed anything related to field overrides and multi-tenancy from the PR.

**Comment by kuipumu** [comment]
@Agrendalath I responded a few of your comments here and in https://github.com/open-craft/openedx-ltistore/pull/2#discussion_r1316968655. According to what you previously commented on [the openedx-ltistore PR](https://github.com/open-craft/openedx-ltistore/pull/2#discussion_r1311840228), I reviewed again the possibility of keep using the keyset/token views from this app instead of creating a new views on the openedx-ltistore app, I was able to do so, I removed the views from the openedx-ltistore PR and modified the keyset/token views on xblock-lti-consumer.

**Comment by Agrendalath** [comment]
Thanks, @kuipumu! I will review and test both PRs tomorrow.

File: lti_consumer/static/js/xblock_studio_view.js
Comment (Agrendalath) [inline lti_consumer/static/js/xblock_studio_view.js:75]: [We are not adding field overrides in this PR](https://github.com/openedx/xblock-lti-consumer/pull/390#discussion_r1316979821), so I believe we should remove the changes from this file.

File: lti_consumer/tests/unit/plugin/test_views.py
Comment (Agrendalath) [inline lti_consumer/tests/unit/plugin/test_views.py:778]: Typo: `nonexistent`. It seems to be occurring in more places in this file.

File: lti_consumer/plugin/views.py
Comment (Agrendalath) [inline lti_consumer/plugin/views.py:450]: I'm thinking about this again and wonder if we actually need to pass the external ID to these views. Wouldn't using the `lti_config_id` (UUID) instead be better? Whenever we use the external configuration, at least one `LtiConfiguration` entry always points there.

In other words:
1. We change `lti_consumer/api.py` to provide the UUID instead of the `namespace/config_id` in the URLs.
2. In this view, we use `lti_config = LtiConfiguration.objects.get(config_id=lti_config_id)` to instantiate the config from the DB.
3. We run `lti_config.get_lti_consumer()` and still get the correct `LtiConsumer` because [this code](https://github.com/openedx/xblock-lti-consumer/blob/f15a78c6054f34411b0a4610083cba91fe5dad7c/lti_consumer/models.py#L567-L584) retrieves the configuration from the external store.

File: lti_consumer/models.py
Comment (Agrendalath) [reply lti_consumer/models.py]: @kuipumu, Quince will be released [in December](https://docs.google.com/spreadsheets/d/11DheEtMDGrbA9hsUvZ2SEd4Cc8CaC4mAfoV8SVaLBGI/edit#gid=195838733&range=AV5:BA5).

**Comment by Agrendalath** [comment]
@kuipumu, please see [this comment](https://github.com/openedx/xblock-lti-consumer/pull/390#discussion_r1320813187) first, as it can affect other comments.

File: lti_consumer/plugin/views.py
Comment (kuipumu) [reply lti_consumer/plugin/views.py:450]: What you are suggesting is doing the same has when the lti_config_id parameter is received on the public keyset and access token view but using the external configuration values?. The only issue I think about this is that on the `lti_consumer/api.py` [get_lti_1p3_launch_info](https://github.com/openedx/xblock-lti-consumer/blob/f15a78c6054f34411b0a4610083cba91fe5dad7c/lti_consumer/api.py#L112) function the returned `token_url` and `keyset_url` will be pointing to the current `LtiConfiguration` UUID, in that case a tool deployment for each LTI consumer XBlock will also need to be created, defeating in some way the use of an external LTI configuration, unless `get_lti_1p3_launch_info` returns a `token_url` and `keyset_url` of a common `LtiConfiguration` for all `LtiConfiguration` instances using the same external configuration, I don't think this last approach to be OK for get_lti_1p3_launch_info. Please correct me if I'm wrong on what I understood about your suggestion or the issue I raised is incorrect.

File: lti_consumer/plugin/views.py
Comment (kuipumu) [reply lti_consumer/plugin/views.py]: I created a ExternalConfigurationNotFound exception and raised that instead, I applied this on the latest commit.

File: lti_consumer/plugin/urls.py
Comment (kuipumu) [reply lti_consumer/plugin/urls.py]: I thought the app name could be less strict, just replaced str with slug since the app name should be a slug to, just applied this fix on the latest commit.

File: lti_consumer/models.py
Comment (kuipumu) [reply lti_consumer/models.py:394]: Yes, I forgot to update these methods once I removed the use of value overrides, I also updated the LTI store PR to include these missing fields on the external config.

File: lti_consumer/lti_xblock.py
Comment (kuipumu) [reply lti_consumer/lti_xblock.py:701]: I added a validation to the external ID format on the latest commit.

File: lti_consumer/api.py
Comment (kuipumu) [reply lti_consumer/api.py]: I applied this suggestion on the latest commit, I also added the comment on this alternative format in the URL definition.

File: lti_consumer/api.py
Comment (kuipumu) [reply lti_consumer/api.py]: Thanks, I added a validation to external_id on the LtiConfiguration clean method on the last commit.

File: lti_consumer/static/js/xblock_studio_view.js
Comment (kuipumu) [reply lti_consumer/static/js/xblock_studio_view.js]: I modified this file to remove all LTI 1.3 fields except external ID and modified this comment on the latest commit.

File: lti_consumer/tests/unit/plugin/test_views.py
Comment (kuipumu) [reply lti_consumer/tests/unit/plugin/test_views.py]: I fixed all occurrences of this typo on the tests in this file on the latest commit.

**Comment by kuipumu** [comment]
Hi @Agrendalath , I applied all your suggestions except for [this previous comment](https://github.com/openedx/xblock-lti-consumer/pull/390#discussion_r1320813187) which I just left a reply, I also responded and applied your suggestions and responded your comments on https://github.com/open-craft/openedx-ltistore/pull/2

**Comment by Agrendalath** [comment]
@kuipumu, thank you; I will review this and https://github.com/open-craft/openedx-ltistore/pull/2 tomorrow.

File: lti_consumer/lti_xblock.py
Comment (Agrendalath) [inline lti_consumer/lti_xblock.py:1704]: Something odd happens when I try to use the external configuration directly from the XBlock. When I set the `Configuration Type` to `Reusable Configuration` and provide a valid `LTI Reusable Configuration ID`, the XBlock break with `Error: (1048, "Column 'version' cannot be null")`. The relevant part of the traceback is:
```python
File "/edx/shared-src/xblock-lti-consumer/lti_consumer/lti_xblock.py", line 1177, in author_view
  return self.student_view(context)
File "/edx/shared-src/xblock-lti-consumer/lti_consumer/lti_xblock.py", line 1227, in student_view
  context.update(self._get_context_for_template())
File "/edx/shared-src/xblock-lti-consumer/lti_consumer/lti_xblock.py", line 1730, in _get_context_for_template
  lti_consumer = self._get_lti_consumer()
File "/edx/shared-src/xblock-lti-consumer/lti_consumer/lti_xblock.py", line 1128, in _get_lti_consumer
  return get_lti_consumer(config_id_for_block(self))
File "/edx/shared-src/xblock-lti-consumer/lti_consumer/api.py", line 96, in config_id_for_block
  config = _get_lti_config_for_block(block)
File "/edx/shared-src/xblock-lti-consumer/lti_consumer/api.py", line 76, in _get_lti_config_for_block
  lti_config = _get_or_create_local_lti_config(
File "/edx/shared-src/xblock-lti-consumer/lti_consumer/api.py", line 46, in _get_or_create_local_lti_config
  lti_config.save()
File "/edx/shared-src/xblock-lti-consumer/lti_consumer/models.py", line 313, in save
  super().save(*args, **kwargs)
```

**Review by Agrendalath — COMMENTED** [review]
@kuipumu, I'm having some trouble with testing it. I explained one issue in the comment. The other one is that when I set the configuration to `CONFIG_EXTERNAL` in the `LtiConfiguration` DB entry, it gets reverted back to `CONFIG_ON_DB` every time I load the XBlock.

I tried setting the external ID to `lti_store:1` and `lti_store:test-configuration` (slug), and some other permutations, but none seem to be working for me.

I already enabled the following flags: `lti_consumer.enable_database_config`, `lti_consumer.enable_external_config_filter`. Did I miss anything in my setup?

File: lti_consumer/lti_xblock.py
Comment (kuipumu) [reply lti_consumer/lti_xblock.py:1699]: This is because the external ID set on the XBlock is not found, this is either because the reusable configuration feature is not properly setup or the external configuration doesn't exist. If you check this [specific line](https://github.com/openedx/xblock-lti-consumer/blob/d315f5afb6f01c0efc4534dab7ba88bcf9bb0ce2/lti_consumer/api.py#L77) you can notice that if no config is found the "version" value will return None, once this value is sent to the `_get_or_create_local_lti_config` function, when it tries to save the new values to the `LtiConfiguration` it fails since "version" value cannot be null. This could be fixed by adding an extra validation to any required field from the external config on `_get_lti_config_for_block` but I think it's out of the scope of this PR, an issue about this should be created.

File: lti_consumer/models.py
Comment (kuipumu) [reply lti_consumer/models.py]: I modified this validation to use the regex instead, it should work even if the value is of external_id is None

File: lti_consumer/models.py
Comment (kuipumu) [reply lti_consumer/models.py]: Sure, I just added an external_config cached_property.

File: lti_consumer/lti_xblock.py
Comment (kuipumu) [reply lti_consumer/lti_xblock.py:703]: The regex check will raise TypeError if data.external_config is None, but since the value is being transformed into str that will not happen, and the regex validation will fail to validate a string with a value of 'None'. I just applied your suggestion to mix both validations in a single condition and also to improve the validation on the model.

**Comment by kuipumu** [comment]
@Agrendalath in response to [this comment](https://github.com/openedx/xblock-lti-consumer/pull/390#pullrequestreview-1652575040).

This issue is happening both for configurations using LTI 1.1 and LTI 1.3, It seems that once you load the studio view for the XBlock, the LTI configuration reverts back to the values set on the XBlock, I'm not sure how these values are sync between the XBlock and the model, will need to check further to address this issue, I would think this should be addressed in another PR since it's not directly related to the use of external configurations with LTI 1.3.

I followed the steps from this PR to setup the external reusable configuration: https://github.com/openedx/xblock-lti-consumer/pull/239 

Did you properly setup the `OPEN_EDX_FILTERS_CONFIG` setting?, you should have it setup on the LMS and CMS settings with the following:

```
OPEN_EDX_FILTERS_CONFIG = {
    # LTI Consumer External Configuration Store.
    "org.openedx.xblock.lti_consumer.configuration.listed.v1": {
        "fail_silently": False,
        "pipeline": [
            "lti_store.pipelines.GetLtiConfigurations",
        ],
    },
}
```

**Comment by kuipumu** [comment]
@Agrendalath I just responded to your previous comments and applied your suggestions, I also included an ADR on how to setup this feature for both LTI 1.1 and LTI 1.3.

**Comment by kuipumu** [comment]
@Agrendalath I just applied your suggestions related to the setup instructions and fixed the sync_configurations method, I also extended the README.md of the openedx-ltistore to include the instructions on how to use the "Filter Key" on this PR https://github.com/open-craft/openedx-ltistore/pull/2

File: README.rst
Comment (Agrendalath) [inline README.rst:300]:
```suggestion
1. Install and setup the `openedx-ltistore`_ plugin on the LMS and Studio.
2. Go to LMS admin -> WAFFLE_UTILS -> Waffle flag course override
   (http://localhost:18000/admin/waffle_utils/waffleflagcourseoverridemodel/).
3. Create a waffle flag course override with these values:
    - Waffle flag: lti_consumer.enable_external_config_filter
    - Course id: <your course id>
    - Override choice: Force On
    - Enabled: True
4. Create a new external LTI configuration and use it in the XBlock.
   This is explained in the README of the `openedx-ltistore`_ repository.

.. _openedx-ltistore: https://github.com/open-craft/openedx-ltistore
```

**Review by Agrendalath — APPROVED** [review]
@kuipumu, thank you so much for your persistence!
Before we merge this, please update the version to `9.7.0` and add an entry to the `CHANGELOG`.

:+1:

* [x] I tested this: LTI 1.3 configurations can be reused
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation

File: README.rst
Comment (Agrendalath) [reply README.rst:300]: GitHub does not highlight the changes for me, so I'll add this manually.

Changed lines: 290, 291, 292, 298, 299.
Added lines: 300, 301.

File: lti_consumer/lti_xblock.py
Comment (Agrendalath) [reply lti_consumer/lti_xblock.py:1699]: Created #420.

**Comment by kuipumu** [comment]
@Agrendalath Just applied your latest suggestions, I also update the package version and added a new entry to the CHANGELOG

**Comment by openedx-webhooks** [comment]
@kuipumu 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/xblock-lti-consumer&entry.867055334=https://github.com/openedx/xblock-lti-consumer/pull/390&entry.1484655318=https://github.com/kuipumu&entry.752974735=2023-07-10+22:34&entry.1917517419=2023-10-24+16:46&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
